src/index.rs

```diff
@@ -1,65 +1,90 @@
 use std::collections::{HashMap, HashSet};
 use std::fs::{self, File, Metadata};
-use std::io::{BufRead, BufReader, Write};
-use std::ops::Add;
+use std::io::BufReader;
+use std::io::BufWriter;
 use std::path::{Path, PathBuf};
-use std::str::FromStr;
-use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use std::time::{Duration, SystemTime};
 use walkdir::{DirEntry, WalkDir};
 
-use log;
+use crate::{id::ResourceId, ArklibError, Result, ARK_FOLDER, INDEX_PATH};
 
-use crate::id::ResourceId;
-use crate::{ArklibError, Result, ARK_FOLDER, INDEX_PATH};
+pub const RESOURCE_UPDATED_THRESHOLD: Duration = Duration::from_millis(1);
+pub type Paths = HashSet<PathBuf>;
 
-#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug)]
+/// IndexEntry represents a [`ResourceId`] and the time it was last modified
+#[derive(
+    Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug, Serialize, Deserialize,
+)]
 pub struct IndexEntry {
+    /// The time the resource was last modified
     pub modified: SystemTime,
+    /// The resource's ID
     pub id: ResourceId,
 }
 
-#[derive(PartialEq, Clone, Debug)]
+/// Represents an index of resources in the system
+///
+/// This struct maintains mappings between resource IDs and their corresponding
+/// file paths, as well as mappings between file paths and index entries
+/// Additionally, it keeps track of collisions that occur during indexing
+#[serde_as]
+#[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
 pub struct ResourceIndex {
-    pub id2path: HashMap<ResourceId, CanonicalPathBuf>,
-    pub path2id: HashMap<CanonicalPathBuf, IndexEntry>,
-
+    /// A mapping of resource IDs to their corresponding file paths
+    #[serde_as(as = "Vec<(_, _)>")]
+    pub id2path: HashMap<ResourceId, PathBuf>,
+    /// A mapping of file paths to their corresponding index entries
+    pub path2id: HashMap<PathBuf, IndexEntry>,
+    /// A mapping of resource IDs to the number of collisions they have
     pub collisions: HashMap<ResourceId, usize>,
+    /// The root path of the index
     root: PathBuf,
 }
 
+/// Represents an update to the resource index
+///
+/// This struct holds information about resources that have been deleted
+/// or added during an update operation on the resource index
 #[derive(PartialEq, Debug)]
 pub struct IndexUpdate {
+    /// Set of resource IDs that have been deleted
     pub deleted: HashSet<ResourceId>,
-    pub added: HashMap<CanonicalPathBuf, ResourceId>,
+    /// Map of file paths to resource IDs that have been added
+    pub added: HashMap<PathBuf, ResourceId>,
 }
 
-pub const RESOURCE_UPDATED_THRESHOLD: Duration = Duration::from_millis(1);
-
-pub type Paths = HashSet<CanonicalPathBuf>;
-
 impl ResourceIndex {
+    /// Returns the number of entries in the index
+    ///
+    /// Note that the actual size is lower in presence of collisions
     pub fn size(&self) -> usize {
-        //the actual size is lower in presence of collisions
         self.path2id.len()
     }
```

Comment:
Do you think separate functions is a good idea?
1. `pub fn count_files() -> usize` which equals to `self.path2id.len()`
2. `pub fn count_resources() -> usize` which equals to `self.id2path.len()`

Reply:
Probably we should make the mappings private.

Reply:
Updated the maps of `ResourceIndex` to be private. 
I also added 2 methods, `ResourceIndex::count_files()` and `ResourceIndex::count_resources()` to return the size of `path2id` and `id2path` respectively